source/lib.ts

```diff
@@ -0,0 +1,271 @@
+	if (
+		typeof options.store.increment !== 'function' ||
+		typeof options.store.resetKey !== 'function' ||
+		(options.skipFailedRequests &&
+			typeof options.store.decrement !== 'function')
+	) {
+		throw new Error(
+			'An invalid store was passed. Please ensure that the store is a class that implements the `Store` interface.',
+		)
+	}
+
+	// Throw an error if any deprecated options are passed
+	const deprecatedOptions = ['global', 'delayMs', 'delayAfter'] as Array<
+		keyof Options
+	>
+	for (const option of deprecatedOptions) {
+		// This doesn't trigger if any value is set to a falsy value (e.g., 0),
+		// because that essentially disables them
+		if (passedOptions[option]) {
+			throw new Error(
+				`The \`${option}\` option is deprecated and will likely be removed from the \`express-rate-limit\` package in the future.`,
+			)
+		}
+	}
+
+	// Return the 'clean' options
+	return options
+}
+
+/**
+ * Just pass on any errors for the developer to handle, usually as a HTTP 500
+ * Internal Server Error.
+ *
+ * @param fn {Express.RequestHandler} - The request handler for which to handle errors
+ *
+ * @returns {Express.RequestHandler} - The request handler wrapped with a `.catch` clause
+ *
+ * @private
+ */
+const handleAsyncErrors =
+	(fn: Express.RequestHandler): Express.RequestHandler =>
+	async (
+		request: Express.Request,
+		response: Express.Response,
+		next: Express.NextFunction,
+	) => {
+		try {
+			// eslint-disable-next-line @typescript-eslint/no-confusing-void-expression
+			await Promise.resolve(fn(request, response, next)).catch(next)
+		} catch (error: unknown) {
+			next(error)
+		}
+	}
+
+/**
+ *
+ * Create an instance of IP rate-limiting middleware for Express.
+ *
+ * @param passedOptions {Options} - Options to configure the rate limiter
+ *
+ * @returns {RateLimitRequestHandler} - The middleware that rate-limits clients based on your configuration
+ *
+ * @public
+ */
+const rateLimit = (
+	passedOptions?: Partial<Options>,
+): RateLimitRequestHandler => {
+	// Parse the options and add the default values for unspecified options
+	const options = parseOptions(passedOptions ?? {})
+
+	// Then return the actual middleware
+	const middleware = handleAsyncErrors(
+		async (
+			request: Express.Request,
+			response: Express.Response,
+			next: Express.NextFunction,
+		) => {
+			// First check if we should skip the request
+			const skip = await Promise.resolve(options.skip(request, response))
+			if (skip) {
+				next()
+				return
+			}
+
+			// Create an augmented request
+			const augmentedRequest = request as AugmentedRequest
+
+			// Get a unique key for the client
+			const key = await Promise.resolve(options.keyGenerator(request, response))
+			// Increment the client's hit counter by one
+			const { totalHits, resetTime } = await new Promise<{
+				totalHits: number
+				resetTime: Date | undefined
+			}>((resolve, reject) => {
+				options.store.increment(
```

Comment:
I like the rename to `increment`, but we also need to support `options.store.incr()` here for backwards compatibility with all of the third-party stores.